src/simulator/particle/LinkedCellContainer.cpp

```diff
@@ -0,0 +1,372 @@
+}
+
+bool LinkedCellContainer::is_within_domain(
+    const std::array<double, 3> &position) {
+  return position[0] >= left_corner_coordinates[0] &&
+         position[0] < left_corner_coordinates[0] + domain_size_[0] &&
+         position[1] >= left_corner_coordinates[1] &&
+         position[1] < left_corner_coordinates[1] + domain_size_[1] &&
+         position[2] >= left_corner_coordinates[2] &&
+         (domain_size_.size() == 3
+              ? position[2] < left_corner_coordinates[2] + domain_size_[2]
+              : position[2] < left_corner_coordinates[2] + r_cutoff_);
+}
+
+void LinkedCellContainer::update_particle_location(
+    ParticlePointer p, const std::array<double, 3> &old_position) {
+  if (is_within_domain(old_position)) {
+    size_t i = static_cast<size_t>(
+        (old_position[0] - left_corner_coordinates[0]) / r_cutoff_);
+    size_t j = static_cast<size_t>(
+        (old_position[1] - left_corner_coordinates[1]) / r_cutoff_);
+    size_t k =
+        domain_size_.size() == 3
+            ? static_cast<size_t>(
+                  (old_position[2] - left_corner_coordinates[2]) / r_cutoff_)
+            : 0;
+    size_t old_index = i + j * x + k * x * y;
+    cells[old_index].particles.erase(
+        std::remove(cells[old_index].particles.begin(),
+                    cells[old_index].particles.end(), p),
+        cells[old_index].particles.end());
+  }
+  if (is_within_domain(p->getX())) {
+    size_t i = static_cast<size_t>((p->getX()[0] - left_corner_coordinates[0]) /
+                                   r_cutoff_);
+    size_t j = static_cast<size_t>((p->getX()[1] - left_corner_coordinates[1]) /
+                                   r_cutoff_);
+    size_t k =
+        domain_size_.size() == 3
+            ? static_cast<size_t>((p->getX()[2] - left_corner_coordinates[2]) /
+                                  r_cutoff_)
+            : 0;
+    size_t index = i + j * x + k * x * y;
+    cells[index].particles.push_back(p);
+  } else {
+    p->left_domain = true;
+  }
+}
+
+Cell &LinkedCellContainer::get_cell(size_t index) { return cells[index]; }
+
+std::vector<ParticlePointer> LinkedCellContainer::get_neighbours(Particle &p) {
+  std::array<double, 3> position = p.getX();
+  size_t i = static_cast<size_t>((position[0] - left_corner_coordinates[0]) /
+                                 r_cutoff_);
+  size_t j = static_cast<size_t>((position[1] - left_corner_coordinates[1]) /
+                                 r_cutoff_);
+  size_t k = domain_size_.size() == 3
+                 ? static_cast<size_t>(
+                       (position[2] - left_corner_coordinates[2]) / r_cutoff_)
+                 : 0;
+
+  int index = i + j * x + k * x * y;
+  std::vector<ParticlePointer> neighbours;
+
+  int X = static_cast<int>(x);
+  int Y = static_cast<int>(y);
+  int Z = static_cast<int>(z);
+  // handle 2d case
+  // Convert 1D index to 3D coordinates
+  i = index / (Y * Z);
+  j = (index % (Y * Z)) / Z;
+  k = index % Z;
```

Comment:
this feels a bit weird. why did we start with `i,j,k` then calculate `index` and then again calculate `i,j,k`?